### PR TITLE
Make audio mixing robust against lagging sources

### PR DESCRIPTION
**Pull request by marcan**
### Description

This change makes the OBS mixing tree robust against sources with lagging audio. When a source lags beyond the ability to be synced by the audio buffering, only that source will be affected, not the entire mix. df4eb82 fixed a pathological bug that made this happen persistently for a given source and caused OBS audio output to completely fall over; this change addresses the underlying issue that a single misbehaving source breaks global audio output, transiently or not.

### Motivation and Context

Historically, `warning: Max audio buffering reached!` has been treated as some kind of inherent deeply-seated problem; once you get that message all bets are off and restarting OBS is the only solution. This is despite the fact that there are many situations where getting that message does *not* result in persistent problems in practice.

This is not a deep design problem. It's just a bug.

OBS has the ability to buffer audio in order to time-align sources with slightly different audio timestamps. This buffer can grow, up to a maximum of approximately one second, which introduces latency to the stream output. If sources slip in time beyond this maximum, they cannot be reconciled, and a dropout must occur. However, this situation need not and should not result in any catastrophic failure, nor should it affect audio output globally. It should only affect the misbehaving source.

The problem is as follows: when a source has a timestamp that is in the past (before the current audio tick), this triggers OBS to elide mixing, grow its buffer, and try again later. However, this can no longer happen once the buffer is maxed out. At that point, mixing proceeds and audio is processed with the timestamp in the past. Unfortunately, OBS's mixing tree stages cannot cope with out of range timestamps. Scene and transition mixing nodes will pick the lowest input timestamp and treat that as the base for the current audio tick, thus mixing audio for the wrong span of time. Then, `mix_audio` in obs-audio.c will notice the out-of-range incoming timestamp, prior to the current audio tick -which at this point is coming from a top-level transition source- and refuse to mix the audio entirely. This generates a global audio drop-out. At this point, latter code will usually notice the misbehaving source and kick it in some way, but it is too late, as we've already generated a drop-out. And if this sticks for some reason - as it did prior to df4eb82 - then audio breaks forever.

The solution is to catch these lagging sources immediately after their audio is rendered, in the case where the global audio buffering time has reached the limit. At this point we drop their audio partially or fully, and mark the source as pending if necessary, so that latter stages in the mixing tree never see out-of-range timestamps in an active audio input. We also log an error so the user knows that something was up. This creates a drop-out for that source (which can also happen via other codepaths when a source is lagging anyway), but it most importantly avoids any damage to anything *else* that is feeding audio to the mixing tree.

### How Has This Been Tested?

Set up two RTMP media sources, and feed them from two instances of ffmpeg via some RTMP reflector (e.g. nginx-rtmp), such as:

```ffmpeg -stream_loop 100 -re -y -i BigBuckBunny.mp4 -c copy -f flv rtmp://<stream url>```

Set up the audio faders such that the first instance is mixed into the stream, but the second instance is muted. Start streaming (or recording) and view the stream output (or recording).

Use ^Z to pause the second instance, wait a few seconds, and resume it (`fg`). This should trigger the OBS audio buffering to instantly max out. Now repeat the process a few times. Without this patch, each pause will often create a drop-out in stream output, affecting the audio from the first (untouched) instance. With this patch, the audio output will be completely clean, with no drop-outs, as the second source is muted. Unmuting source 2 and continuing the test shows that it drops out as expected while the first source continues on cleanly.

System: Gentoo Linux / amd64

### Types of changes
- Bug fix (non-breaking change which fixes an issue)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.

### CODE REVIEW

**Comment by marcan** [comment]
Improvements in the force-push:

* Do not attempt to ignore_audio for sources with a custom audio render function (instead complain and assert for debug builds; this shouldn't happen)
* Properly lock the audio buffer mutex around ignore_audio
* Recalculate audio output buffer if ignore_audio recovered the laggy source
* When ignore_audio decides a source is way too far out of sync, trigger a reset of the timing code in obs-source.c, which might fix the timestamps.

**Comment by jp9000** [comment]
I greatly appreciate the thorough, in-depth explanation. It makes me feel much more comfortable when seeing code in something as delicate as obs-audio.c. I normally worry greatly when I see people modifying it to a significant extent.

I lament the design of the audio buffering. It had good intent: to try to minimize audio latency as much as humanly possible, which it usually succeeds in doing, but the complexity required to do due to the hierarchical audio mixing made it a troublesome thing in hindsight. I wish I had just used a fixed latency value from the beginning, or at least a more simple method of dynamic audio buffering somehow or another. I'm not necessarily going to call it a case of trying to be "clever" all that much, but it's definitely proof that sometimes it's better to accept some sacrifices in order to keep code complexity (and potential bugs from that complexity) down.

**Comment by marcan** [comment]
Yeah, while reading the code I got the feeling that it was *trying* to do things a certain way which had good intentions, but ended up resulting in a bunch of problems that required workarounds, and eventually the code has become very hard to follow and reason about.

I'd say it's worth putting revisiting it and rewriting a bunch of this code somewhere on the radar (whether you ever get there and do it... that's another question, as there are only so many hours in a day).

I'm certainly not 100% certain that the patch included here is fully correct, but I wrote it from the point of view of being defensive and at least I'm pretty confident that, given the current design, cleaning up or disabling sources right after render time if something goes wrong is the right thing to do, to keep the rest of the code from falling over. I'm less certain about the recovery path for those sources; the idea of resetting the timing stuff in obs-source.c makes sense in principle, and definitely helped for the pathological case of bad JACK timestamps from #3413, but I certainly can't prove to myself that it won't result in a pathologic loop of resets under some conditions or with some sources. At least, though, it'll still only affect just the one problem source, not the rest.

**Comment by jp9000** [comment]
Do you think it would be okay to make those `LOG_DEBUG` messages guarded by `DEBUG_AUDIO == 1`? Almost every source gets logged by those messages and it makes the debug output very spammy.
